Firmware/Filament_sensor.cpp

```diff
@@ -282,14 +277,7 @@ void IR_sensor_analog::voltUpdate(uint16_t raw) { // to be called from the ADC I
 }
 
 uint16_t IR_sensor_analog::getVoltRaw() {
-    uint16_t newVoltRaw;
-    ATOMIC_BLOCK(ATOMIC_RESTORESTATE) { newVoltRaw = voltRaw; }
-    return newVoltRaw;
-}
-
-void IR_sensor_analog::settings_init() {
-    IR_sensor::settings_init();
-    sensorRevision = (SensorRevision)eeprom_read_byte((uint8_t *)EEPROM_FSENSOR_PCB);
+    ATOMIC_BLOCK(ATOMIC_RESTORESTATE) { return voltRaw; }
```

Comment:
This can't be right. Wouldn't this fail to enable the interrupts?

Reply:
this _should_ be right if the author of the block wrote it properly

Reply:
It’s a macro for C. I doubt it’s fine. I think they used some do_while shenanigans

Reply:
old code:
```
uint16_t IR_sensor_analog::getVoltRaw() {
    uint16_t newVoltRaw;
    ATOMIC_BLOCK(ATOMIC_RESTORESTATE) { newVoltRaw = voltRaw; }
		in r24, 0x3f ; 63
    return newVoltRaw;

static __inline__ uint8_t __iCliRetVal(void){
    cli();
		cli
		ldd r16, Y+13 ; 0x0d
		ldd r17, Y+14 ; 0x0e
    (void)__s;
}
static __inline__ void __iRestore(const  uint8_t *__s){
    SREG = *__s;
		out 0x3f, r24 ; 63
            uint16_t volt = getVoltRaw();
```

new code:
```
uint16_t IR_sensor_analog::getVoltRaw() {
    ATOMIC_BLOCK(ATOMIC_RESTORESTATE) { return voltRaw; }
		in r24, 0x3f ; 63
static __inline__ uint8_t __iCliRetVal(void){
    cli();
		cli
		ldd r16, Y+13 ; 0x0d
		ldd r17, Y+14 ; 0x0e
    (void)__s;
}
static __inline__ void __iRestore(const  uint8_t *__s){
    SREG = *__s;
		out 0x3f, r24 ; 63
            uint16_t volt = getVoltRaw();
```

Reply:
https://blog.oddbit.com/post/2019-02-01-atomicblock-magic-in-avrlibc/
_ATOMIC_BLOCK creates a block of code that is guaranteed to be executed atomically. Upon entering the block the Global Interrupt Status flag in SREG is disabled, and re-enabled upon exiting the block **from any exit path**._

what the compiler does:
https://godbolt.org/z/B_y4c3